server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -46,49 +45,38 @@
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Objects;
+import java.util.function.Function;
 import java.util.stream.Stream;
 
 
 public class DocumentMapper implements ToXContentFragment {
 
     public static class Builder {
-
-        private final Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers = new LinkedHashMap<>();
+        private final Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper> metadataMappers;
         private final RootObjectMapper rootObjectMapper;
         private final Mapper.BuilderContext builderContext;
         private final IndexSettings indexSettings;
         private final IndexAnalyzers indexAnalyzers;
-        private final DocumentMapperParser documentMapperParser;
         private final DocumentParser documentParser;
 
         private Map<String, Object> meta;
 
         public Builder(RootObjectMapper.Builder builder, MapperService mapperService) {
-            this.indexSettings = mapperService.getIndexSettings();
-            this.indexAnalyzers = mapperService.getIndexAnalyzers();
-            this.documentMapperParser = mapperService.documentMapperParser();
-            this.documentParser = mapperService.documentParser();
+            this(builder, mapperService.getIndexSettings(), mapperService.getIndexAnalyzers(), mapperService.documentParser(),
+                mapperService::getMetadataMappers);
+        }
+
+        Builder(RootObjectMapper.Builder builder,
+                IndexSettings indexSettings,
+                IndexAnalyzers indexAnalyzers,
+                DocumentParser documentParser,
+                Function<String, Map<Class<? extends MetadataFieldMapper>, MetadataFieldMapper>> metadataMappersFunction) {
+            this.indexSettings = indexSettings;
+            this.indexAnalyzers = indexAnalyzers;
+            this.documentParser = documentParser;
             this.builderContext = new Mapper.BuilderContext(indexSettings.getSettings(), new ContentPath(1));
             this.rootObjectMapper = builder.build(builderContext);
-            final String type = rootObjectMapper.name();
-            final DocumentMapper existingMapper = mapperService.documentMapper(type);
```

Comment:
@romseygeek as part of the backport, due to this call to documentMapper that takes a type, I had to make this argument a function, and the corresponding method in MapperService now takes a String argument. Is that good enough or can you thin of further simplications?

Reply:
I think it fine; it's not beautiful, but that's types for you.